staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1264,31 +1274,36 @@ func (c *Cacher) getBookmarkAfterResourceVersionLockedFunc(ctx context.Context,
 //   - start at Any (return the current watchCache's RV)
 //
 // The returned function must be called under the watchCache lock.
-func (c *Cacher) getStartResourceVersionForWatchLockedFunc(ctx context.Context, parsedWatchResourceVersion uint64, opts storage.ListOptions) (func() uint64, error) {
+func (c *Cacher) getStartResourceVersionForWatchLockedFunc(parsedResourceVersion, watchCacheResourceVersion uint64, opts storage.ListOptions) (func() uint64, error) {
 	if opts.SendInitialEvents == nil || *opts.SendInitialEvents {
-		return func() uint64 { return parsedWatchResourceVersion }, nil
+		return func() uint64 { return parsedResourceVersion }, nil
+	}
+	if parsedResourceVersion == 0 {
+		// here we assume that watchCache locked is already held
+		return func() uint64 { return c.watchCache.resourceVersion }, nil
 	}
-	return c.getCommonResourceVersionLockedFunc(ctx, parsedWatchResourceVersion, opts)
+	return func() uint64 { return watchCacheResourceVersion }, nil
 }
 
-// getCommonResourceVersionLockedFunc a helper that simply computes a ResourceVersion
-// based on the input parameters. Please examine callers of this method to get more context.
+// getWatchCacheResourceVersion returns a ResourceVersion to which the watch cache must be synchronized to
 //
-// The returned function must be called under the watchCache lock.
-func (c *Cacher) getCommonResourceVersionLockedFunc(ctx context.Context, parsedWatchResourceVersion uint64, opts storage.ListOptions) (func() uint64, error) {
-	switch {
-	case len(opts.ResourceVersion) == 0:
-		rv, err := storage.GetCurrentResourceVersionFromStorage(ctx, c.storage, c.newListFunc, c.resourcePrefix, c.objectType.String())
-		if err != nil {
-			return nil, err
-		}
-		return func() uint64 { return rv }, nil
-	case parsedWatchResourceVersion == 0:
-		// here we assume that watchCache locked is already held
-		return func() uint64 { return c.watchCache.resourceVersion }, nil
-	default:
-		return func() uint64 { return parsedWatchResourceVersion }, nil
+// Depending on the input parameters, the semantics of the returned ResourceVersion are:
+//   - must be at Exact RV (when parsedWatchResourceVersion > 0)
+//   - can be at Any RV (when parsedWatchResourceVersion = 0)
+//   - must be at Most Recent RV (return an RV from etcd)
+//
+// note that the above semantic is enforced by the API validation (defined elsewhere):
+//
+//	if SendInitiaEvents != nil => ResourceVersionMatch = NotOlderThan
+//	if ResourceVersionmatch != nil => ResourceVersionMatch = NotOlderThan & SendInitialEvents != nil
+//
+// to satisfy the legacy case (SendInitialEvents = true, RV="") we skip checking opts.Predicate.AllowWatchBookmarks
+func (c *Cacher) getWatchCacheResourceVersion(ctx context.Context, parsedWatchResourceVersion uint64, opts storage.ListOptions) (uint64, error) {
+	if opts.SendInitialEvents == nil || !*opts.SendInitialEvents || len(opts.ResourceVersion) != 0 {
```

Comment:
Thanks - I think this works now.
But we really need unit tests for all these 3 function - I would really like to see a full matrix tested in unit tests for all these three functions.

Reply:
These 3 functions should be tested implicitly by other unit tests we have but I don't mind adding explicits tests for testing the input and the output or do you have different tests in mind ?

Reply:
Most of cases are implicitly tested, but I would really like to see a full test matrix for those functions:
all possible combinations of: 
- SendInitialEvents (nil, true, false)
- AllowedWatchBookmarks (true, false)
- ResourceVersion ("", 0, "105")

Reply:
I assume that `105` is a future RV, right ? In that case would it be okay to extend `TestWaitUntilWatchCacheFreshAndForceAllEvents` ?

I think that the other cases are covered by https://github.com/kubernetes/kubernetes/blob/0e238c4c28f11d1ec7cf5f506c81195c1cc463fe/staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go#L1238 and https://github.com/kubernetes/kubernetes/blob/0e238c4c28f11d1ec7cf5f506c81195c1cc463fe/staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go#L1514

Reply:
We probably need both current and future RV.

No - I want a simple unit test testing only this function and nothing else.

Reply:
ah, ok, only `getWatchCacheResourceVersion` ? Do you want to have a similar test for `getBookmarkAfterResourceVersionLockedFunc` and `getStartResourceVersionForWatchLockedFunc` ?

Reply:
It would definitely be desired too - given these are relatively tricky and have simple input/output.